File: keevo/backend/src/main/java/com/keevo/inventory/counting/domain/service/InventoryReportTextFormatter.java
package com.keevo.inventory.counting.domain.service;

import com.keevo.inventory.counting.domain.model.InventoryGapReport;
import com.keevo.inventory.counting.domain.model.InventoryGapRow;
import org.springframework.stereotype.Component;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

/**
 * InventoryReportTextFormatter — GoF Strategy for generating text reports.
 *
 * <p>Generates emoji-rich WhatsApp text and detailed export text,
 * matching the day-close report style (Story 4.4).
 * Story 6.3 — Gap Analysis Report.
 */
@Component
public class InventoryReportTextFormatter {

    private static final ZoneId WAT_ZONE = ZoneId.of("Africa/Lagos");
    private static final DateTimeFormatter FR_DATE =
            DateTimeFormatter.ofPattern("d MMMM yyyy", Locale.FRENCH);
    private static final DateTimeFormatter TIME_FORMAT =
            DateTimeFormatter.ofPattern("HH'h'mm");

    /**
     * Generates emoji-rich compact WhatsApp text.
     * Max 5 "Top manques" entries to keep message compact.
     */
    public String formatWhatsApp(InventoryGapReport report, String actorName) {
        StringBuilder sb = new StringBuilder();
        ZonedDateTime zdt = report.getGeneratedAt().atZone(WAT_ZONE);
        String date = zdt.format(FR_DATE);
        String time = zdt.format(TIME_FORMAT);

        sb.append("📋 Rapport d'inventaire — ").append(report.getStoreName()).append("\n");
        sb.append("📅 ").append(date).append(" — ").append(time).append("\n");
        sb.append("👤 ").append(actorName).append("\n\n");

        sb.append("✅ Concordants : ").append(report.getSummary().totalConcordant())
                .append(" produit").append(report.getSummary().totalConcordant() > 1 ? "s" : "").append("\n");

        if (report.getSummary().totalSurplus() > 0) {
            sb.append("⚠️ Surplus : ").append(report.getSummary().totalSurplus())
                    .append(" produit").append(report.getSummary().totalSurplus() > 1 ? "s" : "")
                    .append(" (+").append(formatXaf(report.getSummary().totalSurplusValueXaf()))
                    .append(")\n");
        }

        if (report.getSummary().totalShortage() > 0) {
            sb.append("🔴 Manquants : ").append(report.getSummary().totalShortage())
                    .append(" produit").append(report.getSummary().totalShortage() > 1 ? "s" : "")
                    .append(" (−").append(formatXaf(report.getSummary().totalShortageValueXaf()))
                    .append(")\n");
        }

        if (!report.getShortageRows().isEmpty()) {
            sb.append("\nTop manques :\n");
            report.getShortageRows().stream().limit(5).forEach(row -> {
                sb.append("• ").append(row.productName());
                if (row.variantLabel() != null) sb.append(" ").append(row.variantLabel());
                sb.append(" : ").append(row.ecart()).append(" unité")
                        .append(Math.abs(row.ecart()) > 1 ? "s" : "")
                        .append(" (−").append(formatXaf(row.gapValueXaf())).append(")\n");
            });
        }

        return sb.toString();
    }

    /**
     * Generates detailed text export (for "Télécharger" on Free plan).
     */
    public String formatDetailedText(InventoryGapReport report, String actorName) {
        StringBuilder sb = new StringBuilder();
        ZonedDateTime zdt = report.getGeneratedAt().atZone(WAT_ZONE);
        String date = zdt.format(FR_DATE);
        String time = zdt.format(TIME_FORMAT);

        sb.append("Rapport d'inventaire — ").append(report.getStoreName()).append("\n");
        sb.append("Date : ").append(date).append(" — ").append(time).append("\n");
        sb.append("Réalisé par : ").append(actorName).append("\n\n");

        sb.append("Résumé :\n");
        sb.append("✅ Concordants : ").append(report.getSummary().totalConcordant()).append(" produits\n");

        if (report.getSummary().totalSurplus() > 0) {
            sb.append("⚠️ Surplus : ").append(report.getSummary().totalSurplus())
                    .append(" produits (+").append(formatXaf(report.getSummary().totalSurplusValueXaf()))
                    .append(")\n");
        }
        if (report.getSummary().totalShortage() > 0) {
            sb.append("🔴 Manquants : ").append(report.getSummary().totalShortage())
                    .append(" produits (−").append(formatXaf(report.getSummary().totalShortageValueXaf()))
                    .append(")\n");
        }

        if (!report.getShortageRows().isEmpty()) {
            sb.append("\nDétail des manquants :\n");
            for (InventoryGapRow row : report.getShortageRows()) {
                sb.append("• ").append(row.productName());
                if (row.variantLabel() != null) sb.append(" ").append(row.variantLabel());
                sb.append(" : Keevo ").append(row.theoretical())
                        .append(" → Réel ").append(row.physical())
                        .append(" (").append(row.ecart()).append(")")
                        .append(" — −").append(formatXaf(row.gapValueXaf())).append("\n");
            }
        }

        if (!report.getSurplusRows().isEmpty()) {
            sb.append("\nDétail des surplus :\n");
            for (InventoryGapRow row : report.getSurplusRows()) {
                sb.append("• ").append(row.productName());
                if (row.variantLabel() != null) sb.append(" ").append(row.variantLabel());
                sb.append(" : Keevo ").append(row.theoretical())
                        .append(" → Réel ").append(row.physical())
                        .append(" (+").append(row.ecart()).append(")")
                        .append(" — +").append(formatXaf(row.gapValueXaf())).append("\n");
            }
        }

        return sb.toString();
    }

    String formatXaf(long amount) {
        String str = String.valueOf(amount);
        StringBuilder buf = new StringBuilder();
        int len = str.length();
        for (int i = 0; i < len; i++) {
            if (i > 0 && (len - i) % 3 == 0) buf.append(' ');
            buf.append(str.charAt(i));
        }
        return buf + " FCFA";
    }
}

File: keevo/app/lib/features/inventory/domain/service/inventory_report_text_formatter.dart
import '../../domain/model/inventory_gap_report_model.dart';

/// Formats the inventory gap report as emoji-rich WhatsApp text.
/// Mirrors InventoryReportTextFormatter.java on the backend.
/// Story 6.3.
String formatWhatsAppReport(
    InventoryGapReportModel report, String actorName) {
  final buf = StringBuffer();
  final now = report.generatedAt;
  final date = '${now.day} ${_frenchMonth(now.month)} ${now.year}';
  final time = '${now.hour}h${now.minute.toString().padLeft(2, '0')}';

  buf.writeln('📋 Rapport d\'inventaire — ${report.storeName}');
  buf.writeln('📅 $date — $time');
  buf.writeln('👤 $actorName');
  buf.writeln();
  buf.writeln('✅ Concordants : ${report.summary.totalConcordant} produits');

  if (report.summary.totalSurplus > 0) {
    buf.writeln('⚠️ Surplus : ${report.summary.totalSurplus} produits '
        '(+${formatXaf(report.summary.totalSurplusValueXaf)})');
  }
  if (report.summary.totalShortage > 0) {
    buf.writeln('🔴 Manquants : ${report.summary.totalShortage} produits '
        '(−${formatXaf(report.summary.totalShortageValueXaf)})');
  }

  if (report.shortageRows.isNotEmpty) {
    buf.writeln();
    buf.writeln('Top manques :');
    for (final row in report.shortageRows.take(5)) {
      final label = row.variantLabel != null
          ? '${row.productName} ${row.variantLabel}'
          : row.productName;
      final units = row.ecart.abs() > 1 ? 'unités' : 'unité';
      buf.writeln(
          '• $label : ${row.ecart} $units (−${formatXaf(row.gapValueXaf)})');
    }
  }

  return buf.toString();
}

/// Formats the inventory gap report as detailed text for export/download.
String formatDetailedTextReport(
    InventoryGapReportModel report, String actorName) {
  final buf = StringBuffer();
  final now = report.generatedAt;
  final date = '${now.day} ${_frenchMonth(now.month)} ${now.year}';
  final time = '${now.hour}h${now.minute.toString().padLeft(2, '0')}';

  buf.writeln('RAPPORT D\'INVENTAIRE DÉTAILLÉ');
  buf.writeln('${'=' * 40}');
  buf.writeln('Boutique : ${report.storeName}');
  buf.writeln('Date : $date — $time');
  buf.writeln('Généré par : $actorName');
  buf.writeln('Périmètre : ${report.scope}');
  buf.writeln();

  buf.writeln('RÉSUMÉ');
  buf.writeln('${'-' * 40}');
  buf.writeln('Total comptés : ${report.summary.totalCounted}');
  buf.writeln('Concordants : ${report.summary.totalConcordant}');
  buf.writeln('Surplus : ${report.summary.totalSurplus} '
      '(+${formatXaf(report.summary.totalSurplusValueXaf)})');
  buf.writeln('Manquants : ${report.summary.totalShortage} '
      '(−${formatXaf(report.summary.totalShortageValueXaf)})');
  buf.writeln();

  if (report.shortageRows.isNotEmpty) {
    buf.writeln('MANQUANTS (par valeur décroissante)');
    buf.writeln('${'-' * 40}');
    for (final row in report.shortageRows) {
      final label = row.variantLabel != null
          ? '${row.productName} ${row.variantLabel}'
          : row.productName;
      buf.writeln(
          '$label | Keevo: ${row.theoretical} → Réel: ${row.physical} | '
          'Écart: ${row.ecart} | −${formatXaf(row.gapValueXaf)}');
    }
    buf.writeln();
  }

  if (report.surplusRows.isNotEmpty) {
    buf.writeln('SURPLUS (par valeur décroissante)');
    buf.writeln('${'-' * 40}');
    for (final row in report.surplusRows) {
      final label = row.variantLabel != null
          ? '${row.productName} ${row.variantLabel}'
          : row.productName;
      buf.writeln(
          '$label | Keevo: ${row.theoretical} → Réel: ${row.physical} | '
          'Écart: +${row.ecart} | +${formatXaf(row.gapValueXaf)}');
    }
    buf.writeln();
  }

  if (report.concordantRows.isNotEmpty) {
    buf.writeln('CONCORDANTS');
    buf.writeln('${'-' * 40}');
    for (final row in report.concordantRows) {
      buf.writeln('${row.productName} | Quantité: ${row.theoretical}');
    }
  }

  return buf.toString();
}

/// Format XAF amount with space as thousands separator.
String formatXaf(int amount) {
  final str = amount.toString();
  final buf = StringBuffer();
  for (var i = 0; i < str.length; i++) {
    if (i > 0 && (str.length - i) % 3 == 0) buf.write(' ');
    buf.write(str[i]);
  }
  return '${buf.toString()} FCFA';
}

String _frenchMonth(int month) {
  const months = [
    '',
    'janvier',
    'février',
    'mars',
    'avril',
    'mai',
    'juin',
    'juillet',
    'août',
    'septembre',
    'octobre',
    'novembre',
    'décembre',
  ];
  return months[month];
}
